src/components/views/rooms/PinnedEventsPanel.js

```diff
@@ -0,0 +1,171 @@
+            if (index !== -1) {
+                pinned.splice(index, 1);
+                MatrixClientPeg.get().sendStateEvent(this.props.mxRoom.roomId, 'm.room.pinned_events', {pinned}, '').then(() => {
+                    if (this.props.onUnpinned) this.props.onUnpinned();
+                });
+            } else if (this.props.onUnpinned) this.props.onUnpinned();
+        }
+    },
+    _canUnpin: function() {
+        return this.props.mxRoom.currentState.mayClientSendStateEvent('m.room.pinned_events', MatrixClientPeg.get());
+    },
+    render: function() {
+        const MessageEvent = sdk.getComponent("views.messages.MessageEvent");
+        const MemberAvatar = sdk.getComponent("views.avatars.MemberAvatar");
+
+        const sender = this.props.mxRoom.getMember(this.props.mxEvent.getSender());
+        const avatarSize = 40;
+
+        let unpinButton = null;
+        if (this._canUnpin()) {
+            unpinButton = <img src="img/cancel-red.svg" className="mx_PinnedEventTile_unpinButton" width="8" height="8"
+                               onClick={this.onUnpinClicked} alt={_t('Unpin Message')} title={_t('Unpin Message')} />;
+        }
+
+        return (
+            <div className="mx_PinnedEventTile">
+                <div className="mx_PinnedEventTile_actions">
+                    <AccessibleButton className="mx_PinnedEventTile_gotoButton mx_textButton" onClick={this.onTileClicked}>
+                        { _t("Jump to message") }
+                    </AccessibleButton>
+                    { unpinButton }
+                </div>
+
+                <MemberAvatar member={sender} width={avatarSize} height={avatarSize} />
+                <span className="mx_PinnedEventTile_sender">
+                    {sender.name}
+                </span>
+                <MessageEvent mxEvent={this.props.mxEvent} className="mx_PinnedEventTile_body" />
+            </div>
+        );
+    }
+});
+
+module.exports = React.createClass({
+    displayName: 'PinnedEventsPanel',
+    propTypes: {
+        // The Room from the js-sdk we're going to show pinned events for
+        room: React.PropTypes.object.isRequired,
+
+        onCancelClick: React.PropTypes.func,
+    },
+
+    getInitialState: function() {
+        return {
+            loading: true,
+        };
+    },
+
+    componentDidMount: function() {
+        this._updatePinnedMessages();
+    },
+
+    _updatePinnedMessages: function() {
+        const pinnedEvents = this.props.room.currentState.getStateEvents("m.room.pinned_events", "");
+        if (!pinnedEvents || !pinnedEvents.getContent().pinned) {
+            this.setState({ loading: false, pinned: [] });
+        } else {
+            const promises = [];
+            const cli = MatrixClientPeg.get();
+
+            pinnedEvents.getContent().pinned.map(eventId => {
+                promises.push(cli.getEventTimeline(this.props.room.getUnfilteredTimelineSet(), eventId, 0).then(timeline => {
```

Comment:
Hmm, I don't think this will necessarily find the event as you'll only get whatever has been loaded into the unfiltered timeline set, ie. someone newly entering a room won't be able to load a pinned event posted longer ago than the brief bit of scrollback their client loads, I think.

Reply:
Watching the network pane, it does a request to the context API. It does try to hit the cache first though.

Reply:
Oh I see - I didn't realise that hit the context API! That sounds like exactly what it should be doing then.